src/VirtualTar.ts

```diff
@@ -0,0 +1,510 @@
+        const waitP = new Promise<void>((resolve) => {
+          this.resolveWaitP = resolve;
+        });
+        await waitP;
+        continue;
+      }
+
+      yield* gen();
+    }
+  }
+
+  /**
+   * Writes a chunk to the internal buffer. If the size of the internal buffer
+   * is larger than or equal to 512 bytes, then the chunks are consumed from
+   * the buffer until the buffer falls below this limit.
+   *
+   * Upon yielding a file or directory token, the relevant data is passed along
+   * to the relevant callback. Note tha the callbacks are queued, so call
+   * {@link settle} to wait for all the pending callbacks to resolve. The end
+   * callback is synchronous, so it is executed immeidately instead of being
+   * queued.
+   *
+   * Only usable when parsing an archive.
+   *
+   * @param chunk a chunk of the (or the entire) binary tar file
+   */
+  public write(chunk: Uint8Array): void {
+    if (this.state !== VirtualTarState.PARSER) {
+      throw new errors.ErrorVirtualTarInvalidState(
+        'VirtualTar is not in parser mode',
+      );
+    }
+
+    // Update the working accumulator
+    this.workingAccumulator = utils.concatUint8Arrays(
+      this.workingAccumulator,
+      chunk,
+    );
+
+    while (this.workingAccumulator.byteLength >= constants.BLOCK_SIZE) {
+      const block = this.workingAccumulator.slice(0, constants.BLOCK_SIZE);
+      this.workingAccumulator = this.workingAccumulator.slice(
+        constants.BLOCK_SIZE,
+      );
+      const token = this.parser.write(block);
+      if (token == null) continue;
+
+      if (token.type === 'header') {
+        // If we have an extended header, then set the working header to the
+        // extended type and continue. Otherwise, if we have a file, then set
+        // the token type to file.
+        if (token.fileType === 'extended') {
+          this.workingTokenType = 'extended';
+          continue;
+        } else if (token.fileType === 'file') {
+          this.workingTokenType = 'file';
+        } else {
+          this.workingTokenType = undefined;
+        }
+
+        // If we have additional metadata, then use it to override token data
+        let filePath = token.filePath;
+        if (this.workingMetadata != null) {
+          filePath = this.workingMetadata.path ?? filePath;
+          this.workingMetadata = undefined;
+        }
+
+        if (token.fileType === 'directory') {
+          const p = this.directoryCallback({
+            type: 'directory',
+            path: filePath,
+            stat: {
+              size: token.fileSize,
+              mode: token.fileMode,
+              mtime: token.fileMtime,
+              uid: token.ownerUid,
+              gid: token.ownerGid,
+              uname: token.ownerUserName,
+              gname: token.ownerGroupName,
+            },
+          });
+          this.callbacks.push(p);
+          continue;
+        } else if (token.fileType === 'file') {
```

Comment:
If we're checking the type of something like this in an if-else chain then its usually better to use a switch case instead.